apps/confidential/confidential_test.go

```diff
@@ -66,3 +75,105 @@ lQFhpxvPgPNBuxVRnsxoH/sLOA==
 		t.Fatalf("TestCertFromPEM: got nil key, want key != nil")
 	}
 }
+
+const token = "fake_token"
+
+var tokenScope = []string{"the_scope"}
+
+func fakeClient() (Client, error) {
+	cred, err := NewCredFromSecret("fake_secret")
+	if err != nil {
+		return Client{}, err
+	}
+	client, err := New("fake_client_id", cred, WithAuthority("https://fake_authority/fake"))
+	if err != nil {
+		return Client{}, err
+	}
+	client.base.Token.AccessTokens = &fake.AccessTokens{
+		AccessToken: accesstokens.TokenResponse{
+			AccessToken:   token,
+			ExpiresOn:     internalTime.DurationTime{T: time.Now().Add(1 * time.Hour)},
+			ExtExpiresOn:  internalTime.DurationTime{T: time.Now().Add(1 * time.Hour)},
+			GrantedScopes: accesstokens.Scopes{Slice: tokenScope},
+		},
+	}
+	client.base.Token.Authority = &fake.Authority{
+		InstanceResp: authority.InstanceDiscoveryResponse{
+			TenantDiscoveryEndpoint: "https://fake_authority/fake/discovery/endpoint",
+			Metadata: []authority.InstanceDiscoveryMetadata{
+				{
+					PreferredNetwork: "fake_authority",
+					PreferredCache:   "fake_cache",
+					Aliases: []string{
+						"fake_authority",
+						"fake_auth",
+						"fk_au",
+					},
+				},
+			},
+			AdditionalFields: map[string]interface{}{
+				"api-version": "2020-02-02",
+			},
+		},
+	}
+	client.base.Token.Resolver = &fake.ResolveEndpoints{
+		Endpoints: authority.NewEndpoints("https://fake_authority/fake/auth",
+			"https://fake_authority/fake/token", "https://fake_authority/fake/jwt", "fake_authority"),
+	}
+	client.base.Token.WSTrust = &fake.WSTrust{}
+	return client, nil
+}
+
+func TestAcquireTokenByCredential(t *testing.T) {
+	client, err := fakeClient()
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = client.AcquireTokenSilent(context.Background(), tokenScope)
+	// first attempt should fail
+	if err == nil {
+		t.Fatal("unexpected nil error from AcquireTokenSilent")
+	}
+	tk, err := client.AcquireTokenByCredential(context.Background(), tokenScope)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if tk.AccessToken != token {
+		t.Fatalf("unexpected access token %s", tk.AccessToken)
+	}
+	// second attempt should return the cached token
+	tk, err = client.AcquireTokenSilent(context.Background(), tokenScope)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if tk.AccessToken != token {
+		t.Fatalf("unexpected access token %s", tk.AccessToken)
+	}
+}
+
+func TestAcquireTokenByAuthCode(t *testing.T) {
+	client, err := fakeClient()
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = client.AcquireTokenSilent(context.Background(), tokenScope)
+	// first attempt should fail
+	if err == nil {
+		t.Fatal("unexpected nil error from AcquireTokenSilent")
+	}
+	tk, err := client.AcquireTokenByAuthCode(context.Background(), "fake_auth_code", tokenScope)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if tk.AccessToken != token {
+		t.Fatalf("unexpected access token %s", tk.AccessToken)
+	}
+	// second attempt should return the cached token
```

Comment:
Since we are doing auth code flow, there will be a user as well. So we would need to pick up an account and do an acquire_token_silent with account

Reply:
I see.  So two questions.  How do I create the account (do I just use the result from `client.Accounts()` and should I be worried that the call succeeded without an account?

Reply:
Yeah, client.Accounts() should work. I am still thinking on the second question. I will have to take a closer look.

Reply:
So it seems like the account isn't looked at when reading a token from the cache, presumably due to [this](https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/dev/apps/internal/base/internal/storage/storage.go#L109).

Unfortunately `Accounts()` returns a `nil` slice.  It's because my fake access token doesn't set the `IDToken` field so as a result [this](https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/dev/apps/internal/base/internal/storage/storage.go#L191) check doesn't happen and no account is added to the cache.

Reply:
Yeah Ideally, there should be `IDToken` returned as well if there is a user. So there is an Account in the cache as well and then that would mean Accounts() wont return a nil.
I believe in that case doing an acquireTokenSilent() with no account will also not give an anything from the cache. - Answering your second question - should I be worried that the call succeeded without an account?

Reply:
That doesn't seem to be the case, at least for this test.  i.e. having no account returns the cached token.  Presumably this is just triggering the same code path as a token acquired from a secret.  I will update the test to form a proper token response for a user.

Reply:
This has been fixed and also uncovered a bug which has also been fixed.